pennylane_sf/gbs.py

```diff
@@ -0,0 +1,246 @@
+        if not analytic and backend != "gaussian":
+            raise ValueError("Only the Gaussian backend is supported in non-analytic mode.")
+
+        super().__init__(wires, analytic=analytic, shots=shots, hbar=hbar)
+        self.cutoff = cutoff_dim
+        self.backend = backend
+        self._params = None
+        self._WAW = None
+
+    @staticmethod
+    def _calculate_WAW(params, A, n_mean):
+        """Calculate the :math:`WAW` matrix.
+
+        Rescales :math:`A` so that when encoded in GBS the mean photon number is equal to
+        ``n_mean``.
+
+        Args:
+            params (array[float]): variable parameters
+            A (array[float]): adjacency matrix
+            n_mean (float): mean number of photons
+
+        Returns:
+            array[float]: the :math`WAW` matrix
+        """
+        A *= rescale(A, n_mean)
+        W = np.diag(np.sqrt(params))
+        return W @ A @ W
+
+    @staticmethod
+    def _calculate_n_mean(A):
+        """Calculate the mean number of photons for an adjacency matrix encoded into GBS.
+
+        Args:
+            A (array[float]): adjacency matrix
+
+        Returns:
+            float: mean number of photons
+        """
+        singular_values = np.linalg.svd(A, compute_uv=False)
+        return np.sum(singular_values ** 2 / (1 - singular_values ** 2))
+
+    # pylint: disable=pointless-statement,expression-not-assigned
+    def apply(self, operation, wires, par):
+        self._params, A, _ = par
+
+        if len(self._params) != self.num_wires:
+            raise ValueError(
+                "The number of variable parameters must be equal to the total number " "of wires."
+            )
+
+        self._WAW = self._calculate_WAW(*par)
+        n_mean_WAW = self._calculate_n_mean(self._WAW)
+
+        op = self._operation_map[operation](self._WAW, mean_photon_per_mode=n_mean_WAW / len(A))
+        op | [self.q[wires.index(i)] for i in wires]
+
+        if not self.analytic:
+            MeasureFock() | [self.q[wires.index(i)] for i in wires]
+
+    def pre_measure(self):
+        self.eng = sf.Engine(self.backend, backend_options={"cutoff_dim": self.cutoff})
+
+        if self.analytic:
+            results = self.eng.run(self.prog)
+        else:
+            # NOTE: currently, only the gaussian backend supports shots > 1
+            results = self.eng.run(self.prog, shots=self.shots)
+
+        self.state = results.state
+        self.samples = results.samples
+
+    def probability(self, wires=None):
+        wires = wires or self.wires
+        wires = Wires(wires)
+
+        if self.analytic:
+            return super().probability(wires=wires)
+
+        N = len(wires)
+        samples = np.take(self.samples, self.wires.indices(wires), axis=1)
+
+        probs = all_fock_probs_pnr(samples)
+        ind = np.indices([self.cutoff] * N).reshape(N, -1).T
+
+        probs = OrderedDict((tuple(i), probs[tuple(i)]) for i in ind)
+        return probs
```

Comment:
@antalszava this is another thing to add to the device ADR - standardize the form of returned probabilities!